File: 0018-4sum/0018-4sum.py
class Solution:
    def fourSum(self,nums,target):
        n=len(nums)
        # Array ko sort kardo
        nums.sort()
        res=[]                         # Resultant 2D array

        # First element fix karo
        for i in range(n-3):
            # Duplicate first element skip karo
            if i>0 and nums[i]==nums[i-1]:
                continue

            # Second element fix karo
            for j in range(i+1,n-2):
                # Duplicate second element skip karo
                if j>i+1 and nums[j]==nums[j-1]:
                    continue

                left=j+1
                right=n-1

                while left<right:
                    sum=nums[i]+nums[j]+nums[left]+nums[right]

                    if sum==target:
                        # Quadruplet mil gaya
                        res.append([
                            nums[i],
                            nums[j],
                            nums[left],
                            nums[right]
                        ])

                        # Duplicate third element skip karo
                        while left<right and nums[left]==nums[left+1]:
                            left+=1

                        # Duplicate fourth element skip karo
                        while left<right and nums[right]==nums[right-1]:
                            right-=1

                        left+=1
                        right-=1

                    elif sum<target:
                        # Sum chota hai
                        left+=1

                    else:
                        # Sum bada hai
                        right-=1

        return res

File: 0018-4sum/0018-4sum.cpp
class Solution {
public:
    vector<vector<int>> fourSum(vector<int>&nums,int target) {
        int n=nums.size();

        // Array ko sort kardo
        sort(nums.begin(),nums.end());

        vector<vector<int>>res;                  // Resultant 2D array

        // First element fix karo
        for(int i=0;i<n-3;i++){
            // Duplicate first element skip karo
            if(i>0 && nums[i]==nums[i-1]){
                continue;
            }

            // Second element fix karo
            for(int j=i+1;j<n-2;j++){
                // Duplicate second element skip karo
                if(j>i+1 && nums[j]==nums[j-1]){
                    continue;
                }
                int left=j+1;                   // Third element
                int right=n-1;                  // Fourth element

                while(left<right){
                    long long sum=(long long)nums[i]+nums[j]+nums[left]+nums[right];
                    if(sum==target){
                        // Quadruplet mil gaya
                        res.push_back({
                            nums[i],
                            nums[j],
                            nums[left],
                            nums[right]
                        });
                        // Duplicate third element skip karo
                        while(left<right && nums[left]==nums[left+1]){
                            left++;
                        }
                        // Duplicate fourth element skip karo
                        while(left<right && nums[right]==nums[right-1]){
                            right--;
                        }
                        left++;
                        right--;
                    }
                    else if(sum<target){
                        // Sum chota hai
                        left++;
                    }
                    else{
                        // Sum bada hai
                        right--;
                    }
                }
            }
        }
        return res;
    }
};